File: 2023/05 If You Give A Seed A Fertilizer/c/one/src/main.c
#include <ctype.h>
#include <stdio.h>
#include <string.h>

#include "utils.h"

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        fprintf(stderr, "Usage: %s <INPUT_FILE>\n", argv[0]);
        return 1;
    }

    FILE* const infile = fopen(argv[1], "r");
    if (infile == NULL)
    {
        fprintf(stderr, "ERROR! Failed to open \"%s\"\n", argv[1]);
        return 1;
    }

    // Verify start of seed line
    int bytesRead = 0;
    fscanf(infile, "seeds: %n", &bytesRead);
    if (bytesRead != 7)
    {
        fputs("ERROR! Failed to read seeds list\n", stderr);
        fclose(infile);
        return 1;
    }

    // Used to make cleanup easier
    int exitStatus = 0;

    DynamicArray seeds;
    dynamicArrayCreate(&seeds, sizeof(unsigned long long int));

    // Read seeds
    for (unsigned long long int seed; fscanf(infile, "%llu", &seed) == 1;)
    {
        dynamicArrayPushBack(&seeds, &seed, 1);
    }

    // Array of bool flags representing if a seed at a given index was covnerted during the current map
    // Reset at start of each map
    DynamicArray seedConverted;
    dynamicArrayCreate(&seedConverted, sizeof(unsigned char));
    dynamicArrayResize(&seedConverted, seeds.size);

    // Read each line

    LineBuffer line;
    lineBufferCreate(&line);

    for (int lineNum = 1; readNextLine(infile, &line); ++lineNum)
    {
        // Skip empty lines
        if (line.size == 0)
            continue;

        // Process mapping if line starts with digit
        if (isdigit(line.line[0]))
        {
            // Read mapping
            unsigned long long int destStart;
            unsigned long long int srcStart;
            unsigned long long int size;
            if (sscanf(line.line, "%llu %llu %llu", &destStart, &srcStart, &size) != 3)
            {
                fprintf(stderr, "ERROR! Failed to read line %d\n", lineNum);
                exitStatus = 1;
                goto cleanup;
            }

            // Try to convert all seeds
            for (size_t i = 0; i < seeds.size; ++i)
            {
                unsigned long long int* const seed = (unsigned long long*)dynamicArrayIndex(&seeds, i);
                unsigned char* const converted = (unsigned char*)dynamicArrayIndex(&seedConverted, i);
                if (!*converted && *seed >= srcStart && *seed < srcStart + size)
                {
                    *converted = true;
                    *seed = destStart + *seed - srcStart;
                }
            }
        }
        // Otherwise a new mapping is starting. Reset conversion flags
        else
        {
            memset(seedConverted.array, 0, sizeof(unsigned char) * seedConverted.size);
        }
    }

    printf("Result: %llu\n", *minull((const unsigned long long int*)seeds.array, seeds.size));

cleanup:
    lineBufferDestroy(&line);

    dynamicArrayDestroy(&seedConverted);
    dynamicArrayDestroy(&seeds);

    fclose(infile);

    return exitStatus;
}

File: 2023/05 If You Give A Seed A Fertilizer/rust/one/src/main.rs
use std::env;
use std::fs;

#[derive(Debug)]
enum ParseSeedsError {
    EmptyLine,
    WrongStart,
}

// Parses seeds from string in format of "seeds: 1 2 3"
fn parse_seeds(string: &str) -> Result<Vec<i64>, ParseSeedsError> {
    let mut tokens = string.split(" ");

    // Make sure at least one token is present
    if let Some(start) = tokens.next() {
        // Verify start of line
        if start == "seeds:" {
            let mut seeds = Vec::new();

            // Parse all found seeds and insert them into seeds
            while let Some(seed) = tokens.next() {
                let seed = seed
                    .trim()
                    .parse()
                    .expect("Unexpcted non-number in seeds list");
                seeds.push(seed);
            }

            Ok(seeds)
        } else {
            Err(ParseSeedsError::WrongStart)
        }
    } else {
        Err(ParseSeedsError::EmptyLine)
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Usage: {} <INPUT_FILE>", args[0]);
    }

    // Read file into lines
    let contents = fs::read_to_string(&args[1]).expect("Failed to read input file");
    let lines: Vec<&str> = contents.lines().collect();

    let mut seeds = parse_seeds(lines[0]).expect("Failed to parse seeds");

    // Flag to check if a seed at a given index has been converted yet by the current map. All set to false on start of every map
    let mut seeds_converted: Vec<bool> = vec![false; seeds.len()];

    // Iterate over each line, skipping seed line, and removing empty lines
    for (line_num, line) in lines
        .iter()
        .enumerate()
        .skip(2)
        .filter(|(_line_num, line)| !line.is_empty())
    {
        // If line starts with digit it's a mapping. Otherwise it's the start of a new map
        if line.chars().nth(0).unwrap().is_ascii_digit() {
            // Parse numbers on line
            let tokens: Vec<i64> = line
                .split(" ")
                .map(|string| {
                    string
                        .trim()
                        .parse::<i64>()
                        .expect("Failed to parse mapping value")
                })
                .collect();

            // There should only be three numbers per mapping
            if tokens.len() != 3 {
                panic!("Invalid number on line {line_num}");
            }

            // Extract three numbers by name
            let dest_start = tokens[0];
            let src_start = tokens[1];
            let size = tokens[2];

            // Try to convert each seed if it hasn't been converted yet this map
            for (seed, converted) in seeds
                .iter_mut()
                .zip(&mut seeds_converted)
                .filter(|(_seed, converted)| !**converted)
            {
                if *seed >= src_start && *seed < src_start + size {
                    *converted = true;
                    *seed = dest_start + *seed - src_start;
                }
            }
        } else {
            // Reset conversion flags for start of new map
            seeds_converted.fill(false);
        }
    }

    println!(
        "Result: {}",
        seeds.iter().min().expect("Must contain at least one seed")
    );
}
